grease/src/Grease/Shape.hs

```diff
@@ -314,3 +342,267 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+  -- if we dont know where to place members we have to fail/just place some bytes
+  buildMember :: HasPtrWidth w => Word64 -> MDwarf.Member -> StateT VisitState Maybe (Word64, Seq.Seq (MemShape w NoTag))
+  buildMember loc member =
+    do
+      memLoc <- lift $ MDwarf.memberLoc member
+      padd <- lift $ Just (if memLoc == loc then Seq.empty else Seq.singleton (padding $ memLoc - loc))
+      memberShapes <- shapeOfTyApp $ MDwarf.memberType member
+      let memByteSize = sum $ memShapeSize ?ptrWidth <$> memberShapes
+      let nextLoc = memLoc + fromIntegral memByteSize
+      lift $ Just (nextLoc, padd Seq.>< memberShapes)
+
+  shapeOfTyApp :: (HasPtrWidth w) => MDwarf.TypeRef -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeOfTyApp x = shapeSeq =<< lift . extractType sprog =<< pure x
+
+  shapeSeq :: (HasPtrWidth w) => MDwarf.TypeApp -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeSeq (MDwarf.UnsignedIntType w) = ishape w
+  shapeSeq (MDwarf.SignedIntType w) = ishape w
+  shapeSeq MDwarf.SignedCharType = ishape (1 :: Int)
+  shapeSeq MDwarf.UnsignedCharType = ishape (1 :: Int)
+  -- TODO(#263): Need modification to DWARF to collect DW_TAG_count and properly evaluate dwarf ops for upper bounds in subranges
+  shapeSeq (MDwarf.ArrayType _elTy _ub) = lift $ Nothing
+  -- need to compute padding between elements, in-front of the first element, and at end
+  -- we could get a missing DWARF member per the format so we have to do something about padding in-front
+  -- even if in practice this should not occur.
+  shapeSeq (MDwarf.StructType sdecl) =
+    let structSize = MDwarf.structByteSize sdecl
+     in do
+          (endLoc, seqs) <-
+            foldM
+              ( \(currLoc, seqs) mem ->
+                  do
+                    (nextLoc, additionalShapes) <- buildMember currLoc mem
+                    lift $ Just (nextLoc, seqs Seq.>< additionalShapes)
+              )
+              (0, Seq.empty)
+              (MDwarf.structMembers sdecl)
+          let endPad = if endLoc >= structSize then Seq.empty else Seq.singleton (padding $ structSize - endLoc)
+          lift $ Just $ (seqs Seq.>< endPad)
+  shapeSeq (MDwarf.PointerType _ maybeRef) =
+    let mshape = constructPtrMemShapeFromRef tyUnrollBound sprog =<< lift maybeRef
+     in Seq.singleton <$> mshape
+  shapeSeq _ = lift $ Nothing
+
+type VisitState = Map.Map MDwarf.TypeRef Int
+
+nullPtr :: HasPtrWidth w => (MemShape w NoTag)
+nullPtr =
+  let zbyt = TaggedByte{taggedByteTag = NoTag, taggedByteValue = 0}
+   in Exactly (take (fromInteger $ CT.intValue ?ptrWidth `div` 8) $ repeat $ zbyt)
+
+constructPtrMemShapeFromRef :: HasPtrWidth w => TypeUnrollingBound -> Subprogram -> MDwarf.TypeRef -> StateT VisitState Maybe (MemShape w NoTag)
+constructPtrMemShapeFromRef bound@(TypeUnrollingBound tyUnrollBound) sprog ref =
+  do
+    mp <- get
+    let ct = fromMaybe 0 $ Map.lookup ref mp
+    _ <- put (Map.insert ref (ct + 1) mp)
```

Comment:
I still think that this would make a nice helper function (perhaps in a `where` clause) but nbd